op-node-derivers.md

```diff
@@ -0,0 +1,554 @@
+2. Document existing triggers/conditions/effects, grouped by (sub)deriver. (no breaking changes!)
+3. Introduce new derivers for Alt-DA and Interop etc.
+
+Conditions/Effects to specify (grouped by resource here) loosely taken from [specs PR 96](https://github.com/ethereum-optimism/specs/pull/96):
+
+#### Unsafe head
+
+Triggers:
+- Change of unsafe-payloads queue
+- Changed unsafe L2 head
+- Time to sequence (open or seal block)
+- Attributes conflicting with chain
+- Attributes applicable to head
+- (new, maybe) Unsafe-head L1-origin not in or ahead of canonical L1
+- (new, maybe) Sync-status change (if not implied by unsafe L2 head change)
+
+Condition/effect pairs:
+- Unsafe block addition
+- Unsafe block sync-trigger
+- Unsafe block processing
+- Sequencing
+- Payload attributes processing
+
+#### Cross-unsafe head (interop)
+
+Triggers:
+- External dependencies changed
+- Changed unsafe head
+- Changed cross-unsafe head
+
+Condition/effect pairs:
+- Interop safety progression
+- Interop safety reversal
+
+#### Safe head
+
+Triggers:
+- Attributes older than unsafe head
+- (new, maybe) Safe-head L1-origin no longer in canonical L1
+
+Condition/effect pairs:
+- Safety progression
+- Safety reversal
+
+#### Finalized head
+
+Triggers:
+- New L1 finality signal
+
+Condition/effect pairs:
+- Finality progression
+
+#### L1 Input
+
+Triggers:
+- new L1 block
+- invalidated L1 block
+- derivation pipeline not idle
+
+Condition/effect pairs:
+- L1 attributes generation
+- Pipeline reset
+
+
+### Potential for test-vectors
+
+Defining standards for derivers, and making events encodeable as JSON,
+could be a solid basis for multi-client test-vectors.
+A test-runner would simply iterate through a list of encoded events, applied to a deriver,
+on top of some pre-state, and then assert some post-state properties.
+
+### Alt-clients: Rust / Java
+
+Typed events and event-switches are very natural in Rust and Java, and even more common than in Go.
+Alternative rollup node implementations should be able to adopt this pattern.
+
+### Migration path
+
+This one of the more involved refactors, that we cannot execute all at once in op-node.
+This is a proposal for a 3-phase migration:
+
+#### 1) Derivation/Engine decoupling
+
+The derivation pipeline currently is coupled too tightly to the Engine.
+In particular, the `EngineQueue` should be removed,
+to allow the attributes-generation and unsafe-block processing to be truly independent processes.
+
+The derivation is somewhat statefully connected to the engine however:
+- Block-attributes result in later promotion to a "safe" head
+- Not all attributes are equal: those generated from span-batches result in unsafe blocks, until the last in the span is processed.
+- The find-sync-start routine is coupled to the `EngineQueue` `Reset()`
+
+We need a first step that addresses this, so the derivation process that generates the block attributes
+is shareable with the new driver design.
```

Comment:
I like the scope of step 1) Could you elaborate on some details of how this could be done towards the new Deriver design?

Reply:
Will expand the doc here :+1: